### PR TITLE
TPD-1445 Delete Recurring Giving

### PR DESCRIPTION
**Pull request by jason-rice**
fixed issue with redirecting to new giving in profile/giving/contributions
[touchpointsoftware.atlassian.net/browse/TPD-1445](https://touchpointsoftware.atlassian.net/browse/TPD-1445)

### CODE REVIEW

File: CmsWeb/Areas/OnlineReg/Controllers/OnlineRegController.cs [hunk 1]
```diff
@@ -108,7 +108,7 @@ public partial class OnlineRegController : CmsController
             }
 
             string redirectLink = m.org?.RedirectUrl ?? m.masterorg?.RedirectUrl;
-            if (redirectLink.IsEmpty() && m.OnlineGiving() && CurrentDatabase.Setting("UseGivingPages"))
+            if (CurrentDatabase.Setting("UseGivingPages"))
```
Comment (brondavies) [inline CmsWeb/Areas/OnlineReg/Controllers/OnlineRegController.cs:111]: Need to revert this change.  We only want to use dynamic redirection when the org is a former giving org and doesn't have a RedirectUrl

File: CmsWeb/Areas/Finance/Views/FinanceReports/ManagedGiving.cshtml [hunk 2]
```diff
@@ -2,95 +2,95 @@
...
+                        <th>Payment Method</th>
+                        <th>Inactive</th>
+                        <th>Action</th>
+                    </tr>
+                </thead>
+                <tbody>
+                    @foreach (var rg in Model)
+                    {
+                    <tr>
+                        <td>@rg.Name2</td>
```
Comment (josearmandohf) [inline CmsWeb/Areas/Finance/Views/FinanceReports/ManagedGiving.cshtml:33]: Just want to make sure that the <a> element is not necessary here...
<a href="/FinanceReports/ManageGiving2/@rg.PeopleId">@rg.Name2</a>

**Review by josearmandohf — COMMENTED** [review]
Almost there, there are one comment left by Bron and a couple of my own.

**Comment by josearmandohf** [comment]
@jason-rice  CR done, there are one comment left by Bron and a couple of my own.

File: CmsWeb/Areas/OnlineReg/Controllers/OnlineRegController.cs (hunk 1 above)
Comment (jason-rice) [reply CmsWeb/Areas/OnlineReg/Controllers/OnlineRegController.cs:111]: I was instructed by Chris to redirect to the new giving if UseGivingPages setting is true, regardless of whether or not they have a redirect link set up. I can revert this if you want. Should I?

File: CmsWeb/Areas/Finance/Views/FinanceReports/ManagedGiving.cshtml (hunk 2 above)
Comment (brondavies) [reply CmsWeb/Areas/Finance/Views/FinanceReports/ManagedGiving.cshtml:33]: Shouldn't this be `<a href="/FinanceReports/ManageGiving2/@rg.PeopleId">@rg.Name2</a>` like before?

File: CmsWeb/Areas/Finance/Views/FinanceReports/ManagedGiving.cshtml [hunk 3]
```diff
@@ -2,95 +2,95 @@
...
+                        {
+                            <td>
+                                <a class="delete btn btn-sm btn-danger" href="/FinanceReports/DeleteScheduledGift/?scheduledGiftId=@rg.ScheduledGiftId&peopleId=@rg.PeopleId"><i class="fa fa-trash"></i> Delete</a>
+                            </td>
+                        }
+                        else
+                        {
+                            <td>
+                                <a class="delete btn btn-sm btn-danger" href="/Vault/DeleteVaultData/@rg.PeopleId"><i class="fa fa-trash"></i> Delete</a>
+                            </td>
```
Comment (brondavies) [inline CmsWeb/Areas/Finance/Views/FinanceReports/ManagedGiving.cshtml:52]: You can put the `<td></td>` outside of the if statement to be more succinct.

File: CmsWeb/Areas/OnlineReg/Controllers/OnlineRegController.cs (hunk 1 above)
Comment (brondavies) [reply CmsWeb/Areas/OnlineReg/Controllers/OnlineRegController.cs:111]: Yes, this will still satisfy the requirements.  Without those two bits, we would end up redirecting all registrations to giving pages which is not desired.

File: CmsWeb/Areas/People/Controllers/Person/GivingController.cs [hunk 4]
```diff
@@ -144,7 +145,39 @@ public ActionResult ManageGiving()
                 return Redirect("/OnlineReg/" + org);
             }
 
-            return new EmptyResult();
+            return RedirectToAction("RedirectNewGiving");
```
Comment (brondavies) [inline CmsWeb/Areas/People/Controllers/Person/GivingController.cs:148]: You can just `return RedirectNewGiving();` here

File: CmsWeb/Areas/People/Controllers/Person/GivingController.cs [hunk 5]
```diff
@@ -144,7 +145,39 @@ public ActionResult ManageGiving()
...
+        }
+
+        public ActionResult RedirectNewGiving()
+        {
+            if (CurrentDatabase.Setting("UseGivingPages"))
+            {
```
Comment (brondavies) [inline CmsWeb/Areas/People/Controllers/Person/GivingController.cs:154]: This should be checked around line 139 first to skip the org redirect and then use this method.

File: CmsWeb/Areas/People/Controllers/Person/GivingController.cs [hunk 6]
```diff
@@ -144,7 +145,39 @@ public ActionResult ManageGiving()
...
+                        var redirectLink = "/Give/" + givingPage.PageUrl;
+                        return Redirect(redirectLink);
+                    }
+                }
+                else
+                {
+                    var givingPage = (from g in CurrentDatabase.GivingPages where g.DefaultPage == true select g).FirstOrDefault();
+                    var redirectLink = "/Give/" + givingPage.PageUrl;
+                    return Redirect(redirectLink);
+                }
```
Comment (brondavies) [inline CmsWeb/Areas/People/Controllers/Person/GivingController.cs:175]: All of this logic should be moved to a shared component since the OnlineRegController uses the same code.

File: CmsWeb/Areas/People/Controllers/Person/GivingController.cs [hunk 7]
```diff
@@ -180,7 +213,7 @@ public ActionResult OneTimeGift(int? id, int? fundId)
                 return Redirect("/OnlineReg/" + oid);
             }
 
-            return new EmptyResult();
+            return RedirectToAction("RedirectNewGiving");
```
Comment (brondavies) [inline CmsWeb/Areas/People/Controllers/Person/GivingController.cs:216]: Same here, just use `RedirectNewGiving()` but there should be a way to get to the default *one-time* giving page from here since the RedirectNewGiving is only handling recurring giving.

File: UtilityExtensions/Extensions/Dates.cs [hunk 8]
```diff
@@ -377,6 +377,110 @@ private static DateTime GetBuildDate()
             dt = dt.AddHours(TimeZone.CurrentTimeZone.GetUtcOffset(dt).Hours);
             return dt;
         }
+
+        public static string FormatDayOfMonth(int day)
```
Comment (brondavies) [inline UtilityExtensions/Extensions/Dates.cs:381]: There's a more comprehensive, succinct ordinal function here you can use: https://stackoverflow.com/a/9130114/396005

File: CmsWeb/Areas/Finance/Models/FinanceReportsModel.cs [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web;
+
+namespace CmsWeb.Areas.Finance.Models
+{
+    public class FinanceReportsModel : IDbBinder
+    {
+        public FinanceReportsModel() { }
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Finance/Models/FinanceReportsModel.cs:13]: I don't see a warning or anything in VS. Why is this a good idea?

File: CmsWeb/Areas/OnlineReg/Controllers/OnlineRegController.cs (hunk 1 above)
Comment (jason-rice) [reply CmsWeb/Areas/OnlineReg/Controllers/OnlineRegController.cs:111]: Ok, will do!

File: CmsWeb/Areas/Finance/Views/FinanceReports/ManagedGiving.cshtml (hunk 2 above)
Comment (jason-rice) [reply CmsWeb/Areas/Finance/Views/FinanceReports/ManagedGiving.cshtml:33]: Oh I see what you're asking now. I was working on something else so I wasn't sure what you were asking at first Jose. Chris said to remove the details page and change the data displayed in this table instead. That's why I removed this href.

File: CmsWeb/Areas/Finance/Views/FinanceReports/ManagedGiving.cshtml (hunk 3 above)
Comment (jason-rice) [reply CmsWeb/Areas/Finance/Views/FinanceReports/ManagedGiving.cshtml:52]: done

File: CmsWeb/Areas/People/Controllers/Person/GivingController.cs (hunk 4 above)
Comment (jason-rice) [reply CmsWeb/Areas/People/Controllers/Person/GivingController.cs:148]: done

File: CmsWeb/Areas/People/Controllers/Person/GivingController.cs (hunk 5 above)
Comment (jason-rice) [reply CmsWeb/Areas/People/Controllers/Person/GivingController.cs:154]: done

File: CmsWeb/Areas/People/Controllers/Person/GivingController.cs (hunk 7 above)
Comment (jason-rice) [reply CmsWeb/Areas/People/Controllers/Person/GivingController.cs:216]: I added a parameter to handle this.

File: CmsWeb/Areas/People/Controllers/Person/GivingController.cs (hunk 6 above)
Comment (jason-rice) [reply CmsWeb/Areas/People/Controllers/Person/GivingController.cs:175]: done

File: UtilityExtensions/Extensions/Dates.cs (hunk 8 above)
Comment (jason-rice) [reply UtilityExtensions/Extensions/Dates.cs:381]: changed

File: CmsData/Finance/GivingRoutes.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+namespace CmsData.Finance
+{
+    public class GivingRoutes
+    {
+        public static string ReturnGivingPageUrl(CMSDataContext db, bool? oneTimeGiving = false, bool? recurringGiving = false, bool? pledge = false)
+        {
+            if (db.Setting("UseGivingPages"))
+            {
+                var givingType = "";
+                if (db.CurrentUserPerson.CampusId.HasValue)
```
Comment (brondavies) [inline CmsData/Finance/GivingRoutes.cs:17]: Needs a null check on `CurrentUserPerson` in case someone is not logged in.

**Review by brondavies — APPROVED** [review]

File: CmsData/Finance/GivingRoutes.cs (hunk 10 above)
Comment (jason-rice) [reply CmsData/Finance/GivingRoutes.cs]: fixed this!
